### PR TITLE
do_sigver_init: Allow reinitialization of an existing operation.

### LINKED ISSUE #16936: SipHash regression in 3.0.0
Following code produces different results with openssl 1.1.1 and with openssl 3.0.0.  
Basically it calculates SipHash with digest size 8.

1.1.1 produces: 6d 3e 54 c2 2f f1 fe e2
3.0.0 produces: 1e a3 02 13 d5 82 56 89

`  

    uint8_t key[16];   
    memset (key, 1, 16); 
    EVP_PKEY * sipKey = EVP_PKEY_new_raw_private_key (EVP_PKEY_SIPHASH, NULL, key, 16);
    EVP_MD_CTX * mdCtx = EVP_MD_CTX_create ();
    EVP_PKEY_CTX * ctx = NULL;
    EVP_DigestSignInit (mdCtx, &ctx, NULL, NULL, sipKey);
    EVP_PKEY_CTX_ctrl (ctx, -1, EVP_PKEY_OP_SIGNCTX, EVP_PKEY_CTRL_SET_DIGEST_SIZE, 8, NULL);
    EVP_PKEY_free (sipKey);

    uint8_t buf[8], digest[8];
    memset (buf, 0, 8);
    EVP_DigestSignInit (mdCtx, NULL, NULL, NULL, NULL);
    EVP_DigestSignUpdate (mdCtx, buf, 8);
    size_t l = 8;
    EVP_DigestSignFinal (mdCtx, digest, &l);
`

### PR DESCRIPTION
**Pull request by t8m**
Fixes #16936

This is a draft because IMO documentation and some tests need to be added. And perhaps some more fixing/finetunning is also needed.

The question is, do we want this? The current behavior is definitely a regression from 1.1.1 though.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
@levitte @mattcaswell any opinions?

File: crypto/evp/m_sigver.c [hunk 1]
```diff
@@ -66,20 +66,36 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
...
         props = locpctx->propquery;
 
     ERR_set_mark();
 
-    if (evp_pkey_ctx_is_legacy(locpctx))
+    if (evp_pkey_ctx_is_legacy(locpctx)) {
         goto legacy;
+    } else if (reinit) {
+        if (locpctx->operation != (ver ? EVP_PKEY_OP_VERIFYCTX
+                                   : EVP_PKEY_OP_SIGNCTX)
```
Comment (paulidale) [inline crypto/evp/m_sigver.c:88]: indentation

**Review by paulidale — COMMENTED** [review]
Looks like a simple and clear approach.

File: crypto/evp/m_sigver.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/evp/m_sigver.c:88]: Personally I would align the `?` and `:`

**Review by mattcaswell — COMMENTED** [review]
Is it ever valid in 1.1.1 to call EVP_DigestSignInit() multiple times but without all NULL parameters? e.g. to incrementally specify the digest, the key, etc. Or is all NULL the only valid use case?

**Comment by t8m** [comment]
> Is it ever valid in 1.1.1 to call EVP_DigestSignInit() multiple times but without all NULL parameters? e.g. to incrementally specify the digest, the key, etc. Or is all NULL the only valid use case?

Heh, who knows? :grin:

**Comment by t8m** [comment]
> Is it ever valid in 1.1.1 to call EVP_DigestSignInit() multiple times but without all NULL parameters? e.g. to incrementally specify the digest, the key, etc. Or is all NULL the only valid use case?

So my conclusion is that yes, it is possible, but the key has to be specified in the first call. I've changed the patch to do_sigver_init to accommodate for that.

**Comment by t8m** [comment]
ping for reviews

File: test/evp_extra_test.c [hunk 2]
```diff
@@ -1324,6 +1341,57 @@ static int test_EVP_DigestVerifyInit(void)
...
+    EVP_MD_CTX *mdctx = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    int ret = 0;
+    size_t len = 8;
+
+    if (nullprov != NULL)
+        return TEST_skip("Test does not support a non-default library context");
+
+    memset (buf, 0, 8);
+    memset (key, 1, 16);
```
Comment (mattcaswell) [inline test/evp_extra_test.c:1363]: No space after memset

**Review by mattcaswell — APPROVED** [review]
Approved assuming the minor nit is fixed.

File: test/evp_extra_test.c (hunk 2 above)
Comment (t8m) [reply test/evp_extra_test.c]: Fixed

**Comment by t8m** [comment]
Merged to master branch:
commit 3641f04fb06e9679a67da113bab65e5f1bb5e9ba
commit 8cbfc4f67b4e97d423ab4784dbbb54d454c6342a
commit 3ffd23e9529d725903bc97fd45489a77b831876f
commit 816f72d08834ee35ba2615f624b4a29f2717d1c7
commit ae6b68b761b9c5f30897747487ea943ccfab53ba
Cherry picked to openssl-3.0 branch:
commit 487934081d87a0d02bbb9afd6bd650d5d1afe8ea
commit bef4e21fe247753484b26167897afce1dc8cd10c
commit 433c6cab4b4bc187f8120dc2ca0505f81d7b56e6
commit 2e11cb796e1b46159615625c1df34800a1ebae94
commit 33665a9389e216b5a1f64fa48c8795dc78973621

**Pull request closed without merging**